frontend/server/handlers/artifacts.ts

```diff
@@ -239,6 +254,111 @@ function getGCSArtifactHandler(options: { key: string; bucket: string }, peek: n
   };
 }
 
+function getVolumeArtifactsHandler(options: { bucket: string; key: string }, peek: number = 0) {
+  const { key, bucket } = options;
+  return async (req: Request, res: Response) => {
+    let filePath;
+    let volumeMount;
+    try {
+      // get ml-pipeline-ui pod namespace
+      const SERVER_NAMESPACE = serverInfo.getServerNamespace();
+      // get ml-pipeline-ui pod name
+      const POD_NAME = serverInfo.getServerPodName();
+
+      // only check when POD_NAME and serverNamespace can be obtained
+      if (POD_NAME && SERVER_NAMESPACE) {
+        const [pod, err] = await k8sHelper.getPod(POD_NAME, SERVER_NAMESPACE);
+        if (err) {
+          const { message, additionalInfo } = err;
+          console.error(message, additionalInfo);
+          res.status(500).send(message);
+          return;
+        }
+
+        if (!pod) {
+          const message = `Could not get pod ${POD_NAME} in namespace ${SERVER_NAMESPACE}`;
+          res.status(500).send(message);
+          return;
+        }
+
+        // volumes not specified or volume named ${bucket} not specified
+        if (!Array.isArray(pod?.spec?.volumes) || !pod.spec.volumes.find(v => v?.name === bucket)) {
+          const message = `Failed to open volume://${bucket}/${key}, volume ${bucket} not exist`;
+          res.status(400).send(message);
+          return;
+        }
+
+        // find volumes mount
+        if (
+          Array.isArray(pod?.spec?.containers) &&
```

Comment:
I think we should verify retrieved pod yaml matches our expecation upfront, so that we don't need to do many if branches like this again.

You can use the pattern
```
if (something not expected) {
  console.error(xxxxxxxx)
}

if (something else not expected) {
 ...
}

...

at this point everything works, so we don't need to add if statements any more.
```

this approach makes code more readable, because it reads easily in sequence

Reply:
agree!